File: ios/ExpoMusicMetadataReaderModule.swift
import ExpoModulesCore

struct SongData {
  var album: String
}

public class ExpoMusicMetadataReaderModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoMusicMetadataReader')` in JavaScript.
    Name("ExpoMusicMetadataReader")

    // TODO: add functions / events

    // // Defines event names that the module can send to JavaScript.
    // Events("onChange")

    Function("readSongMetadata") { (songUri: String) in
      // TODO: implement Swift logic
      return SongData(album: "TODO")
    }

    // // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    // Function("hello") {
    //   return "Hello world! 👋"
    // }

    // // Defines a JavaScript function that always returns a Promise and whose native code
    // // is by default dispatched on the different thread than the JavaScript runtime runs on.
    // AsyncFunction("setValueAsync") { (value: String) in
    //   // Send an event to JavaScript.
    //   self.sendEvent("onChange", [
    //     "value": value
    //   ])
    // }

    // // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // // view definition: Prop, Events.
    // View(ExpoMusicMetadataReaderView.self) {
    //   // Defines a setter for the `name` prop.
    //   Prop("name") { (view: ExpoMusicMetadataReaderView, prop: String) in
    //     print(prop)
    //   }
    // }
  }
}

File: android/src/main/java/expo/modules/musicmetadatareader/ExpoMusicMetadataReaderModule.kt
package expo.modules.musicmetadatareader
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

import android.media.MediaMetadataRetriever
import android.util.Base64

data class SongData (
  @Field
  val title: String?,
  @Field
  val artist: String?,
  @Field
  val album: String?,
  @Field
  val albumArtist: String?,
  @Field
  val trackNumber: String?,
  @Field
  val totalTracks: String?,
  @Field
  val duration: String?,
  @Field
  val year: String?
) : Record

class ExpoMusicMetadataReaderModule : Module() {
  private fun readSongMetadata(songUri: String): SongData {
    val metadataRetriever = MediaMetadataRetriever()
    metadataRetriever.setDataSource(songUri)

    val title = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_TITLE)
    val artist = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ARTIST)
    val album = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUM)
    val albumArtist = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUMARTIST)
    // TODO: track number / total tracks inaccurate?
    val trackNumber = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_CD_TRACK_NUMBER)
    val totalTracks = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_NUM_TRACKS)
    val duration = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
    // TODO: add date?
    val year = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_YEAR)

    return SongData(title, artist, album, albumArtist, trackNumber, totalTracks, duration, year)
  }

  private fun readSongCoverData(songUri: String): String? {
    val metadataRetriever = MediaMetadataRetriever()
    metadataRetriever.setDataSource(songUri)

    val coverBytes = metadataRetriever.embeddedPicture ?: return null
    return Base64.encodeToString(coverBytes, Base64.DEFAULT)
  }

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoMusicMetadataReader')` in JavaScript.
    Name("ExpoMusicMetadataReader")

    Function("readSongMetadata") { songUri: String ->
      readSongMetadata(songUri)
    }

    Function("readSongCoverData") { songUri: String ->
      readSongCoverData(songUri)
    }
  }
}
